File: python/test/t_Drawable_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott

ot.TESTPREAMBLE()
ot.RandomGenerator.SetSeed(0)

print("Named colors=", ot.Drawable.GetValidColors())
hexa = ot.Drawable.ConvertFromRGB(31, 63, 127)
assert hexa == "#1f3f7f"
#
hexa = ot.Drawable.ConvertFromRGB(0.1, 0.2, 0.3)
assert hexa == "#1a334d"
#
hexa = ot.Drawable.ConvertFromRGBA(31, 63, 127, 191)
assert hexa == "#1f3f7fbf"
#
hexa = ot.Drawable.ConvertFromRGBA(0.1, 0.2, 0.3, 0.4)
assert hexa == "#1a334d66"
#
hexa = ot.Drawable.ConvertFromHSV(215.0, 0.2, 0.3)
assert hexa == "#3d444d"
#
hexa = ot.Drawable.ConvertFromHSVA(215.0, 0.2, 0.3, 0.4)
assert hexa == "#3d444d66"
#
rgb = ot.Drawable.ConvertFromHSVIntoRGB(215.0, 0.2, 0.3)
ott.assert_almost_equal(rgb, [0.24, 0.265, 0.3])
#
hsv = ot.Drawable.ConvertFromRGBIntoHSV(44.0 / 255.0, 160.0 / 255.0, 72.0 / 255.0)
ott.assert_almost_equal(hsv, (134.48275862068965, 0.725, 0.6274509803921569))
#
hsv = ot.Drawable.ConvertFromRGBIntoHSV(44.0 / 255.0, 72.0 / 255.0, 160.0 / 255.0)
ott.assert_almost_equal(hsv, (225.51724137931038, 0.725, 0.6274509803921569))
#
hsv = ot.Drawable.ConvertFromRGBIntoHSV(160.0 / 255.0, 44.0 / 255.0, 72.0 / 255.0)
ott.assert_almost_equal(hsv, (345.51724137931035, 0.725, 0.6274509803921569))
#
print("BuildDefaultPalette=", ot.Drawable.BuildDefaultPalette(10))
print("BuildRainbowPalette=", ot.Drawable.BuildRainbowPalette(10))
print("BuildTableauPalette=", ot.Drawable.BuildTableauPalette(10))
print("BuildDefaultPalette=", ot.Drawable.BuildDefaultPalette(20))
print("BuildTableauPalette=", ot.Drawable.BuildTableauPalette(20))
#
print("Line styles=", ot.Drawable.GetValidLineStyles())
print("Point styles=", ot.Drawable.GetValidPointStyles())
print("Fill styles=", ot.Drawable.GetValidFillStyles())
#
# Get/set the ResourceMap
name = ot.ResourceMap.GetAsString("Drawable-DefaultPaletteName")
assert name == "Tableau"
#
palette = ot.Drawable.BuildDefaultPalette(4)
assert palette == ["#1f77b4", "#ff7f0e", "#2ca02c", "#d62728"]
#
ot.ResourceMap.SetAsString("Drawable-DefaultPaletteName", "Rainbow")
palette = ot.Drawable.BuildDefaultPalette(4)
assert palette == ["#ff0000", "#ccff00", "#00ff66", "#0066ff"]
#
ot.ResourceMap.SetAsString("Drawable-DefaultPaletteName", "Tableau")
palette = ot.Drawable.BuildDefaultPalette(4)
assert palette == ["#1f77b4", "#ff7f0e", "#2ca02c", "#d62728"]

File: lib/test/t_Drawable_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class Drawable for standard methods
 *
 *  Copyright 2005-2024 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  setRandomGenerator();

  try
  {
    fullprint << "Named colors=" << Drawable::GetValidColors() << std::endl;
    fullprint << "RGB colors (31, 63, 127)=" << Drawable::ConvertFromRGB(31UL, 63UL, 127UL) << std::endl;
    fullprint << "RGB colors (0.1, 0.2, 0.3)=" << Drawable::ConvertFromRGB(0.1, 0.2, 0.3) << std::endl;
    fullprint << "RGBA colors (31, 63, 127, 191)=" << Drawable::ConvertFromRGBA(31UL, 63UL, 127UL, 191UL) << std::endl;
    fullprint << "RGBA colors (0.1, 0.2, 0.3, 0.4)=" << Drawable::ConvertFromRGBA(0.1, 0.2, 0.3, 0.4) << std::endl;
    fullprint << "HSV colors (215.0, 0.2, 0.3)=" << Drawable::ConvertFromHSV(215.0, 0.2, 0.3) << std::endl;
    fullprint << "HSVA colors (215.0, 0.2, 0.3, 0.4)=" << Drawable::ConvertFromHSVA(215.0, 0.2, 0.3, 0.4) << std::endl;
    fullprint << "HSV (215.0, 0.2, 0.3) to RGB=" << Drawable::ConvertFromHSVIntoRGB(215.0, 0.2, 0.3) << std::endl;
    fullprint << "RGB (44, 160, 72) to HSV=" << Drawable::ConvertFromRGBIntoHSV(44UL, 160UL, 72UL) << std::endl;
    fullprint << "RGB (44, 72, 160) to HSV=" << Drawable::ConvertFromRGBIntoHSV(44UL, 72UL, 160UL) << std::endl;
    fullprint << "RGB (160, 44, 72) to HSV=" << Drawable::ConvertFromRGBIntoHSV(160UL, 44UL, 72UL) << std::endl;
    fullprint << "RGB (44.0/255.0, 160.0/255.0, 72.0/255.0) to HSV=" << Drawable::ConvertFromRGBIntoHSV(44.0 / 255.0, 160.0 / 255.0, 72.0 / 255.0) << std::endl;
    fullprint << "RGB (44.0/255.0, 72.0/255.0, 160.0/255.0) to HSV=" << Drawable::ConvertFromRGBIntoHSV(44.0 / 255.0, 72.0 / 255.0, 160.0 / 255.0) << std::endl;
    fullprint << "RGB (160.0/255.0, 44.0/255.0, 72.0/255.0) to HSV=" << Drawable::ConvertFromRGBIntoHSV(160.0 / 255.0, 44.0 / 255.0, 72.0 / 255.0) << std::endl;
    fullprint << "Line styles=" << Drawable::GetValidLineStyles() << std::endl;
    fullprint << "Point styles=" << Drawable::GetValidPointStyles() << std::endl;
    fullprint << "Fill styles=" << Drawable::GetValidFillStyles() << std::endl;
    fullprint << "BuildDefaultPalette=" << Drawable::BuildDefaultPalette(10) << std::endl;
    fullprint << "BuildRainbowPalette=" << Drawable::BuildRainbowPalette(10) << std::endl;
    fullprint << "BuildTableauPalette=" << Drawable::BuildTableauPalette(10) << std::endl;
    fullprint << "BuildDefaultPalette=" << Drawable::BuildDefaultPalette(20) << std::endl;
    fullprint << "BuildTableauPalette=" << Drawable::BuildTableauPalette(20) << std::endl;
    fullprint << "BuildViridisPalette=" << Drawable::BuildViridisPalette(20) << std::endl;
    fullprint << "BuildGrayScalePalette=" << Drawable::BuildGrayScalePalette(20) << std::endl;
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }


  return ExitCode::Success;
}
